src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -0,0 +1,91 @@
+package com.atlassian.bitbucket.jenkins.internal.trigger.register;
+
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketPullRequest;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepository;
+
+import javax.inject.Singleton;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+
+@Singleton
+public class PullRequestStoreImpl implements PullRequestStore {
+
+    private final ConcurrentMap<PullRequestStoreImpl.CacheKey, ConcurrentLinkedQueue<BitbucketPullRequest>> pullRequests;
+
+    public PullRequestStoreImpl() {
+        pullRequests = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void addPullRequest(String serverId, BitbucketPullRequest pullRequest) {
+        PullRequestStoreImpl.CacheKey cacheKey =
+                new PullRequestStoreImpl.CacheKey(pullRequest.getToRef().getRepository().getProject().getKey(),
+                pullRequest.getToRef().getRepository().getSlug(), serverId);
+        pullRequests.computeIfAbsent(cacheKey, key -> {
+            return new ConcurrentLinkedQueue<BitbucketPullRequest>();
+            }).add(pullRequest);
+    }
+
+    @Override
+    public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository repository) {
+
+        PullRequestStoreImpl.CacheKey key =
+                new PullRequestStoreImpl.CacheKey(repository.getProjectKey(), repository.getRepositorySlug(), repository.getServerId());
+        return pullRequests.getOrDefault(key, new ConcurrentLinkedQueue<>())
+                .stream()
+                .filter(pullRequest -> pullRequest.getFromRef().getDisplayId().equals(branchName))
+                .findFirst()
+                .isPresent();
+    }
+
+    @Override
+    public ConcurrentMap<PullRequestStoreImpl.CacheKey, ConcurrentLinkedQueue<BitbucketPullRequest>> getPullRequests() {
```

Comment:
As I've mentioned a few times, we cannot do this. This leaks our internal store, and a mutable copy of that. There is nothing stopping us (or anyone else) from calling this method and then just do `.clear()` on the map to remove all data, or add things in.

We've discussed this, so I think you have a good way forward, but for future reference, if you have to return something like this, you must return an immutable version of it.